design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,349 @@
+the new fields:
+
+```yaml
+apiVersion: v1
+baseDomain: test.metalkube.org
+metadata:
+  name: ostest
+platform:
+  baremetal:
+    networkPolicy:
+      ipPools:
+        traffic-1:
+          cidr: 10.10.10.0/24
+          start: 1
+          end: 10
+          exclude: [4, 7]
+        traffic-2:
+          cidr: 10.10.11.0/24
+          start: 1
+          end: 10
+          exclude: [4, 7]
+      capture:
+        address-from-traffic-1: ippool.traffic-1.allocate
+        address-from-traffic-2: ippool.traffic-2.allocate
+      desiredState:
+        interfaces:
+        - name: eth1
+          type: ethernet
+          state: up
+          ipv4:
+            address:
+            - ip: "{{ capture.address-from-foo.ip }}"
+              prefix-length: "{{ capture.address-from-foo.prefix-length }}"
+            dhcp: false
+            enabled: true
+        - name: eth2
+          type: ethernet
+          state: up
+          ipv4:
+            address:
+            - ip: "{{ capture.address-from-bar.ip }}"
+              prefix-length: "{{ capture.address-from-bar.prefix-length }}"
+            dhcp: false
+            enabled: true
+    hosts:
+      - name: openshift-master-0
+        networkData: [static network configuration for master-0]
+      - name: openshift-worker-0
+        networkData: [static network configuration for worker-0]
+```
+
+The new field `platform.baremetal.networkPolicy` will contain the
+content of the day 2 `NodeNetworkConfigurationPolicy` and also the `ipPools`
+since there is not going to be multiple `networkPolicy` fields.
+
+### Implementation Details/Notes/Constraints
+
+#### Day 2
+
+Since kubernetes-nmstate has a distributed architecture each of the handler
+pods will calculate an allacatable ip from the `NodeNetworkIPPool` status and
+retry the process on conflict
+
+Handlers will check NNCEs at Reconcile and if the ip has being already assigned
+it will re-use that and update `NodeNetworkIPPool` status with it, this way the
+`NodeNetworkIPPool` can be reset but IPs are keep.
+
+
+#### Baremetal IPI
+
+The `image-customize-controller` [1] is a golang controller so it would be quite
+easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
+`networkPolicy`.
+
+[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79
+
+### Risks and Mitigations
+
+Overtime from unexpected failure there are risk of having some IPs allocated
+but not used making the number of available IPs smaller and reaching the end
+sooner than expected, for that reseting the `status.allocated-offsets` should
+reconstruct the pool and bring back those IPs.
+
+### Drawbacks
+
+The baremetal IPI Nmstate do a offline NetworkManager keyfile generation so
+there is no network `currentState` that nmpolicy can reference to, so using
+nmpolicy just to reference the ippools may be too overkill.
```

Comment:
Might be just me, but I honestly don't understand this one - what does the keyfile generation have to do with the network current state ?

Reply:
I want to state that nmpolicy is designed to work in conjuntion with `currentState` so snippets from it can be inserted into the `desiredState`. But at baremetal IPI there is no `currentState` so if user try to put a desiredState with captures other than ones related to ippool they will not work.